nixos/modules/services/x11/display-managers/default.nix

```diff
@@ -441,39 +430,60 @@ in
 
           exit 0
         '';
+      writeSessionFile = {sessionName, script, desktopNames}:
+        pkgs.writeTextFile {
+          name = "${sessionName}-xsession";
+          destination = "/share/xsessions/${sessionName}.desktop";
+          # Desktop Entry Specification:
+          # - https://standards.freedesktop.org/desktop-entry-spec/latest/
+          # - https://standards.freedesktop.org/desktop-entry-spec/latest/ar01s06.html
+          text = ''
+            [Desktop Entry]
+            Version=1.0
+            Type=XSession
+            TryExec=${script}
+            Exec=${script}
+            Name=${sessionName}
+            DesktopNames=${desktopNames}
+          '';
+          } // {
+           providedSessions = [ sessionName ];
+          };
       in
         # We will generate every possible pair of WM and DM.
         concatLists (
-            builtins.map
-            ({dm, wm}: let
-              sessionName = "${dm.name}${optionalString (wm.name != "none") ("+" + wm.name)}";
-              script = xsession dm wm;
-              desktopNames = if dm ? desktopNames
-                             then concatStringsSep ";" dm.desktopNames
-                             else sessionName;
-            in
-              optional (dm.name != "none" || wm.name != "none")
-                (pkgs.writeTextFile {
-                  name = "${sessionName}-xsession";
-                  destination = "/share/xsessions/${sessionName}.desktop";
-                  # Desktop Entry Specification:
-                  # - https://standards.freedesktop.org/desktop-entry-spec/latest/
-                  # - https://standards.freedesktop.org/desktop-entry-spec/latest/ar01s06.html
-                  text = ''
-                    [Desktop Entry]
-                    Version=1.0
-                    Type=XSession
-                    TryExec=${script}
-                    Exec=${script}
-                    Name=${sessionName}
-                    DesktopNames=${desktopNames}
-                  '';
-                } // {
-                  providedSessions = [ sessionName ];
-                })
-            )
+          builtins.map
+            ({dm, wm}: optional dm.supportExternalWM
+              (let
+                sessionName = "${dm.name}+${wm.name}";
+                startup = dm.genStart wm.bin;
```

Comment:
shouldn't this be `wm.start` instead of `wm.bin`?

Reply:
my system fails to start from this pr (rebased on master) because the waitPID thing fails (the additional indirection introduced when calling the resulting script, hides the variable from the caller). If I replace this with `wm.start` as it used to be in the old code, the wm startup script gets included verbatim and it works again.

Reply:
@xaverdh could you provide a reproducing configuration?

Reply:
This should do (tested in vm)
```nix
{ pkgs, ... }:
{
  users.users.bob.isNormalUser = true;
  services.xserver = {
    enable = true;
    windowManager.xmonad.enable = true;
    displayManager = {
      autoLogin = {
        enable = true;
        user = "bob";
      };
      lightdm = {
        enable = true;
        greeter.enable = false;
      };
    };
  };
}
```
You can leave out the auto login stuff and log in manually as `bob` if you want.

Reply:
The configuration doesn't work on `master` (without this PR) on QEMU:
```
rm -f nixos.qcow2 && nixos-rebuild build-vm -I nixos-config=./configuration.nix -I nixpkgs=. && ./result/bin/run-nixos-vm
```

Reply:
It works for me from `master`. What error do you get / in what way does it fail?
Note: If you just get a black screen, then everything is working (`xmonad` is a rather minimal window manager).

Reply:
I got a black screen. OK, I proceed to testing the PR.

Reply:
The current implementation generates the following script:

```
#! /nix/store/cwnwyy82wrq53820z6yg7869z8dl5s7g-bash-4.4-p23/bin/bash

# Legacy session script used to construct .desktop files from
# `services.xserver.displayManager.session` entries. Called from
# `sessionWrapper`.

# Start the window manager.
systemd-cat -t xmonad -- /nix/store/4nl2kcgi927xckk4p8lkxpxi4sxd65wy-xmonad-with-packages-8.10.4/bin/xmonad  &
waitPID=$!


# Start the desktop manager.
if [ -e $HOME/.background-image ]; then
  /nix/store/zv8gfabp44k5f1zxl3dkq7pvx6kqvpls-feh-3.6.3/bin/feh --bg-scale  $HOME/.background-image
fi




test -n "$waitPID" && wait "$waitPID"

/run/current-system/systemd/bin/systemctl --user stop graphical-session.target

exit 0
```

The new implementation generates the following script:

```
#!/nix/store/cwnwyy82wrq53820z6yg7869z8dl5s7g-bash-4.4-p23/bin/bash
# Legacy session script used to construct .desktop files from
# `services.xserver.displayManager.session` entries. Called from
# `sessionWrapper`.

# Run the startup script.
/nix/store/71ql8j8sdcnc5dpnn35syvy1h4gjwpv4-run-xmonad
if [ -e $HOME/.background-image ]; then
  /nix/store/zv8gfabp44k5f1zxl3dkq7pvx6kqvpls-feh-3.6.3/bin/feh --bg-scale  $HOME/.background-image
fi




test -n "$waitPID" && wait "$waitPID"

/run/current-system/systemd/bin/systemctl --user stop graphical-session.target

exit 0
```

where `/nix/store/71ql8j8sdcnc5dpnn35syvy1h4gjwpv4-run-xmonad` is the following.

```
#!/nix/store/cwnwyy82wrq53820z6yg7869z8dl5s7g-bash-4.4-p23/bin/bash
systemd-cat -t xmonad -- /nix/store/4nl2kcgi927xckk4p8lkxpxi4sxd65wy-xmonad-with-packages-8.10.4/bin/xmonad  &
waitPID=$!
```

The script sets `waitPID` but it is not propagated to the parent script.

So, using `wm.start` (raw script text) instead of `wm.bin` (script file name), as @xaverdh states, will fix this.

@poscat0x04 could you fix this?